// @crowdlinker/nestjs-mailer/dist/nodemailer-core.module.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var NodemailerCoreModule_1;
Object.defineProperty(exports, "__esModule", { value: true });
const nodemailer_1 = require("./nodemailer");
const constants_1 = require("./constants");
const common_1 = require("@nestjs/common");
let NodemailerCoreModule = NodemailerCoreModule_1 = class NodemailerCoreModule {
    static forRoot(options) {
        return {
            module: NodemailerCoreModule_1,
            providers: [
                {
                    name: constants_1.NODEMAILER_OPTIONS,
                    provide: constants_1.NODEMAILER_OPTIONS,
                    useValue: options,
                },
                nodemailer_1.Nodemailer,
            ],
            exports: [nodemailer_1.Nodemailer],
        };
    }
    static forRootAsync(options) {
        const asyncProviders = this.createAsyncProviders(options);
        return {
            module: NodemailerCoreModule_1,
            imports: options.imports,
            providers: [...asyncProviders, nodemailer_1.Nodemailer],
            exports: [nodemailer_1.Nodemailer],
        };
    }
    static createAsyncProviders(options) {
        if (options.useExisting || options.useFactory) {
            return [this.createAsyncOptionsProvider(options)];
        }
        const useClass = options.useClass;
        return [
            this.createAsyncOptionsProvider(options),
            {
                provide: useClass,
                useClass,
            },
        ];
    }
    static createAsyncOptionsProvider(options) {
        if (options.useFactory) {
            return {
                name: constants_1.NODEMAILER_OPTIONS,
                provide: constants_1.NODEMAILER_OPTIONS,
                useFactory: options.useFactory,
                inject: options.inject || [],
            };
        }
        const inject = [
            (options.useExisting || options.useClass),
        ];
        return {
            name: constants_1.NODEMAILER_OPTIONS,
            provide: constants_1.NODEMAILER_OPTIONS,
            useFactory: (optionsFactory) => __awaiter(this, void 0, void 0, function* () { return yield optionsFactory.createNodemailerOptions(); }),
            inject,
        };
    }
};
NodemailerCoreModule = NodemailerCoreModule_1 = __decorate([
    common_1.Global(),
    common_1.Module({})
], NodemailerCoreModule);
exports.NodemailerCoreModule = NodemailerCoreModule;
